Update README with detailed explanations (#39)
meilisearch/docs-searchbar.js: README.md

The best way to handle your search bar documentation with [MeiliSearch](https://github.com/meilisearch/MeiliSearch).

![docs-searchbar.js example](/assets/docs-searchbar-example.png)

If you use Vuepress for your documentation, you might check out our [Vuepress plugin for MeiliSearch](https://github.com/meilisearch/vuepress-plugin-meilisearch).

## Table of Contents <!-- omit in toc -->

- [Installation and Usage](#installation-and-usage)
  - [Before using the library](#before-using-the-library)
  - [Basic Usage](#basic-usage)
  - [Customization](#customization)
- [Development Workflow](#development-workflow)
- [Credits](#credits)

## Installation and Usage

### Before using the library

> To integrate a search bar to your documentation and use this library, you need to scrap your website contents and to push it into a MeiliSearch instance.

Reviewer: ```suggestion
The goal of this library is to provide a very easy integration of a searchbar to your own documentation. To make that possible, you need to scrape your website content in advance, and index it in a MailiSearch instance.
```

Reviewer 2: Minor changes.
```suggestion
The goal of this library is to provide very easy integration of a search bar into your own documentation. To make that possible, you need to scrape your website content in advance, and index it in a MeiliSearch instance.
```

Author: > Good! Shouldn't it be better to have a contributing.md than putting ewverything in the readme?

I let you put an issue in our repository `integration-guides` 😜

Author: @eskombro and @matboivin are you sure about `to scrape`?
It means "érafler". Could you update your suggestion @matboivin? 🙂

Reviewer: I think this may be handled more in an organization Level as @erlend-sh suggest here:
https://github.com/meilisearch/discuss/discussions/25
So let's move that discussion there

> To integrate a search bar to your documentation and use this library, you need to scrap your website contents and to push it into a MeiliSearch instance.

Reviewer 2: ```suggestion
To integrate a search bar to your documentation and use this library, you need to scrape content from your website and push it into a MeiliSearch instance.
```

> The goal of this library is to provide very easy integration of a search bar into your own documentation. To make that possible, you need to scrape your website content in advance, and index it in a MeiliSearch instance.

Reviewer 3: Scraiping is not the only way to gather your website content. Most multi-language websites have their content already stored in databases or files and thus they do not need to be parsed.

```suggestion
The goal of this library is to provide very easy integration of a search bar into your own documentation. To make that possible, you need to gather your website content in advance, and index it in a MeiliSearch instance.
```

> Hopefully, we provide all the tools needed and [guide you into these different steps](https://docs.meilisearch.com/resources/howtos/search_bar_for_docs.html) 🚀

Reviewer: ```suggestion
Luckily, we provide all the tools that you need, and can help you through the whole process, if you follow this guide 🚀
```

> As a first introduction, you might only want to test this library without connecting it to your website.<br>

Reviewer 2: ```suggestion
As a first step, you might only want to test this library without connecting it to your website.<br>
```

Reviewer 3: Since it is only for discovering purpose, I think introduction is best suited.

You can do it by running the playground provided in this repository:

```bash
$ yarn install
$ yarn serve
```

The data comes from MeiliSearch documentation.<br>

> Type `create an indxe` to live the MeiliSearch experience.

Reviewer: ```suggestion
Type `create an index` to live the MeiliSearch experience.
```

Reviewer 3: I think this typo is on purpose

Reviewer: if it is, it would be better to make it clear in some way, cause it can be easily mistaken for a real typo :)

Author: I want our users to test the typo tolerance, I'm going to make it clearer.

### Basic Usage

```html
<!DOCTYPE html>
<html>
  <head>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/docs-searchbar.js@{version}/dist/cdn/docs-searchbar.min.css"
    />
  </head>

  <body>
    <input type="search" id="search-bar-input" />
    <script src="https://cdn.jsdelivr.net/npm/docs-searchbar.js@{version}/dist/cdn/docs-searchbar.min.js"></script>
    <script>
      docsSearchBar({
        hostUrl: "https://mymeilisearch.com",
        apiKey: "XXX",
        indexUid: "docs",
        inputSelector: "#search-bar-input",
        debug: true // Set debug to true if you want to inspect the dropdown
      });
    </script>
  </body>
</html>
```

The `hostUrl` and the `apiKey` fields are the credentials of your MeiliSearch instance.<br>

> `indexUid` is the index identifier into the scraper have pushed your website contents.<br>

Reviewer: This sentence is not clear for me

Reviewer 3: ```suggestion
`indexUid` is the index identifier into which the scraper has added your website contents.<br>
```

Author: @eskombro what do you think?

Author: And @matboivin 😇

Reviewer 3: I would even like to add a third suggestion in which the scraper is not mandatory 
```suggestion
`indexUid` is the index identifier in your MeiliSearch instance in which your website content is stored.<br>
```

Reviewer: I prefer the 3rd one!

Reviewer 2: Same!

Author: Peeeeuuuurrrfaict (= perfect with the french accent)

> `indexUid` is the index identifier into the scraper have pushed your website contents.<br>

Reviewer 2: Maybe I got it wrong.
```suggestion
`indexUid` is the unique identifier of the index in which your website content will be added.<br>
```

Reviewer 3: Since we are on doc-search the scrapping already took place. So it is not in future tense. Could you suggest upon @eskombro highlight of that same sentence?

Author: @bidoubiwa is right about the tense, at this step, the contents have already been added 🙂
This sentence appears twice, I'm going to ping you Mathilde on this sentence discussion.

Reviewer 2: Yes, let's continue in the thread above. ^^

> `inputSelector` is the CSS identifier of the HTML search input.

Reviewer 3: I have changed `identifier` to `id` as it is the attribute's name.
I have removed the CSS part as it is not a CSS attribute but a HTML attribute that **can** be used by the CSS to give style to a given ID. 
https://www.w3schools.com/tags/att_id.asp
```suggestion
`inputSelector` is the `id` attribute of the HTML search input tag.
```

_You haven't scraped your documentation contents yet? Check out [this tutorial](https://docs.meilisearch.com/resources/howtos/search_bar_for_docs.html)._

**WARNING: We recommend providing the MeiliSearch public key, which is enough to perform searches.<br>

> Read more about [MeiliSearch authentication](/guides/advanced_guides/authentication.md).**

Reviewer 2: Was the internal link intended?
```suggestion
Read more about [MeiliSearch authentication](https://docs.meilisearch.com/guides/advanced_guides/authentication.html).**
```

Author: Hou! Good job!

> _You haven't scraped your documentation contents yet? Check out [this tutorial](https://docs.meilisearch.com/resources/howtos/search_bar_for_docs.html)._

Reviewer 3: ```suggestion
_Your documentation content is not indexed yet? Check out [this scraping tutorial] (https://docs.meilisearch.com/resources/howtos/search_bar_for_docs.html). _
```

### Customization

The default behavior of this library fits perfectly for a documentation search bar, but you might need some customizations.

- [Optional parameters](#optional-parameters-) (when calling `docsSearchBar` method)
- [Styling](#styling-) (with CSS)

#### Optional parameters <!-- omit in toc -->

When calling the `docsSearchBar` method, you can add optional fields:

##### `queryHook` <!-- omit in toc -->

> This method is called on every keystroke to transform the typed keywords before querying MeiliSearch. By default, it does not do anything.

Reviewer 2: I changed to `It doesn't do anything by default.` since it is used below in `transformData`.
```suggestion
This method is called on every keystroke to transform the typed keywords before querying MeiliSearch. It doesn't do anything by default.
```

##### `transformData` <!-- omit in toc -->

> This method will be called on every hit before displaying them. It doesn't do anything by default.

Reviewer 2: I changed to `By default, it does not do anything.` since it is used in `queryHook`.
```suggestion
This method will be called on every hit before displaying them. By default, it does not do anything.
```

> This method is called on every keystroke to transform the typed keywords before querying MeiliSearch. By default, it does not do anything.

Reviewer 3: Given the sentence above: 
> When calling the `docsSearchBar` method, you can add optional fields:

I don't understand how `queyHook` is a field and which value I should give him. Do you mean something like this?
```suggestion
queryHook takes a callback function as value. This function will be called on every keystroke to transform the typed keywords before querying MeiliSearch. By default, it does not do anything.
```

The same comment is given to all the other subtitles.

> This method will be called on every hit before displaying them. By default, it does not do anything.

Reviewer 3: ```suggestion
`TransformData` takes a callback function as value. This function will be called on every hit before displaying them. By default, it does not do anything.
```

> `TransformData` takes a callback function as value. This function will be called on every hit before displaying them. By default, it does not do anything.

Reviewer 3: ```suggestion
`transformData` takes a callback function as value. This function will be called on every hit before displaying them. By default, it does not do anything.
```

##### `handleSelected` <!-- omit in toc -->

> This method is called when a suggestion is selected (either from a click or a keystroke). By default, it displays anchor links to the results page. Here is an example to override this behavior:

Reviewer 3: ```suggestion
`handleSelected` takes a callback function a value. This function is called when a suggestion is selected (either from a click or a keystroke). By default, it displays anchor links to the results page. Here is an example to override this behavior:
```

```javascript
docsSearchBar({
  // ...
  handleSelected: function(input, event, suggestion, datasetNumber, context) {
    // Prevents the default behavior on click and rather opens the suggestion
    // in a new tab.
    if (context.selectionMethod === "click") {
      input.setVal("");

      const windowReference = window.open(suggestion.url, "_blank");
      windowReference.focus();
    }
  }
});
```

Note that, by default, you can already open a new tab thanks to the CMD/CTRL + Click action.

> The method is called with the following arguments:

Reviewer 3: ```suggestion
The function is called with the following arguments:
```

- `input`: a reference to the search input element. It comes with the `.open()`, `.close()`, `.getVal()` and `.setVal()` methods.

- `event`: the actual event triggering the selection.

- `suggestion`: the object representing the current selection. It contains a `.url` key representing the destination.

- `datasetNumber`: this should always be equal to 1 as `docs-searchbar.js` is searching into one dataset at a time. You can ignore this attribute.

> - `context`: additional information about the selection. Contains a `.selectionMethod` key that can be either `click`, `enterKey`, `tabKey` or `blur`, depending how the suggestion was selected.

Reviewer 2: ```suggestion
- `context`: additional information about the selection. Contains a `.selectionMethod` key that can be either `click`, `enterKey`, `tabKey` or `blur`, depending on how the suggestion was selected.
```

##### `meilisearchOptions` <!-- omit in toc -->

You can forward search parameters to the MeiliSearch API by using the `meilisearchOptions` key. Checkout out the [MeiliSearch documentation about search parameters](https://docs.meilisearch.com/guides/advanced_guides/search_parameters.html#search-parameters).

For example, you might want to increase the number of results displayed in the dropdown:

```javascript
docsSearchBar({
  meilsearchOptions: {
    limit: 10
  }
});
```

> ##### More Examples <!-- omit in toc -->

Reviewer 3: ```suggestion
#### More Examples <!-- omit in toc -->
```

Author: Same

Here is a basic [HTML file](scripts/playground.html) used in the playground of this repository.

As a more concrete example, you can [check out the configuration](https://github.com/meilisearch/vuepress-plugin-meilisearch/blob/bf2210e131eb5630f632d514f35710ee6576ec7a/MeiliSearchBox.vue#L64) applied in the MeiliSearch plugin for Vuepress.

> #### Styling <!-- omit in toc -->

Reviewer 3: ```suggestion
##### Styling <!-- omit in toc -->
```

Author: There is a misunderstanding. The `More examples` section is a sub-title of `Optional Parameters` section. But `Styling` is not, it's the same level than `Optional Parameters`.

```css
/* Main dropdown wrapper */
.meilisearch-autocomplete .dsb-dropdown-menu {
  width: 500px;
}

/* Main category */
.meilisearch-autocomplete .docs-searchbar-suggestion--category-header {
  color: darkgray;
  border: 1px solid gray;
}

/* Category */
.meilisearch-autocomplete .docs-searchbar-suggestion--subcategory-column {
  color: gray;
}

/* Title */
.meilisearch-autocomplete .docs-searchbar-suggestion--title {
  font-weight: bold;
  color: black;
}

/* Description */
.meilisearch-autocomplete .docs-searchbar-suggestion--text {
  font-size: 0.8rem;
  color: gray;
}

/* Highlighted text */
.meilisearch-autocomplete .docs-searchbar-suggestion--highlight {
  color: blue;
}
```

Here is the [CSS customization](https://github.com/meilisearch/vuepress-plugin-meilisearch/blob/bf2210e131eb5630f632d514f35710ee6576ec7a/MeiliSearchBox.vue#L77-L177) applied in the MeiliSearch plugin for Vuepress.

> **TIPS: To inspect the dropdown markup with your browser tools, you should add `debug: true` to your `docsSearchBar` call to prevent it from closing on inspection.**

Reviewer 2: ```suggestion
**TIPS: When inspecting the dropdown markup with your browser tools, you should add `debug: true` to your `docsSearchBar` call to prevent it from closing on inspection.**
```

> Here is the [CSS customization](https://github.com/meilisearch/vuepress-plugin-meilisearch/blob/bf2210e131eb5630f632d514f35710ee6576ec7a/MeiliSearchBox.vue#L77-L177) applied in the MeiliSearch plugin for Vuepress.

Reviewer 3: What is the purpose of this information? To inform the user they can customize their CSS?

Author: Provide a concrete example with CSS customization for the search bar because it's impossible to detail all the possibilities in the README.

Author: @bidoubiwa do you think it needs to be more explicit?

Reviewer: Oui, en effet c'est dans la section "more examples", ca me parait clair!

Reviewer 3: In which case the "more examples" section should not be ##### and "styling" #### but the opposite !

Author: It's more The `More examples` section is for the first part of the customization (`Optional parameters`), not for the styling part.

But I can do a `More examples` section for the `Styling` section as well 🙂

Reviewer 3: Ah oke! It's good as it is then no problem

## Development Workflow

If you want to contribute, this section describes the steps to follow.

Thank you for your interest in a MeiliSearch tool! ♥️

### Requirements <!-- omit in toc -->

To run this project, you will need:

- Node.js >= v8.7.0, use nvm - [installation instructions](https://github.com/creationix/nvm#install-script)
- Yarn

### Install <!-- omit in toc -->

Install all the dependencies:

```bash
$ yarn install
```

### Build <!-- omit in toc -->

```bash
$ yarn build
```

> This command will build all files in `./dist`. This includes regular and minified files for `<script>` inclusion, as well as classes for `import`ing.

Reviewer: ```suggestion
This command will build all files in `./dist`. This includes regular and minified files for `<script>` inclusion, as well as classes `import`.
```

### Basic Usage

```html
<!DOCTYPE html>
<html>
  <head>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/docs-searchbar.js@{version}/dist/cdn/docs-searchbar.min.css"
    />
  </head>

  <body>
    <input type="search" id="search-bar-input" />
    <script src="https://cdn.jsdelivr.net/npm/docs-searchbar.js@{version}/dist/cdn/docs-searchbar.min.js"></script>
    <script>
      docsSearchBar({
        hostUrl: "https://mymeilisearch.com",
        apiKey: "XXX",
        indexUid: "docs",
        inputSelector: "#search-bar-input",
        debug: true // Set debug to true if you want to inspect the dropdown
      });
    </script>
  </body>
</html>
```

The `hostUrl` and the `apiKey` fields are the credentials of your MeiliSearch instance.<br>
`indexUid` is the index identifier into the scraper have pushed your website contents.<br>
`inputSelector` is the `id` attribute of the HTML search input tag.

_Your documentation content is not indexed yet? Check out [this tutorial](https://docs.meilisearch.com/resources/howtos/search_bar_for_docs.html)._

> **WARNING: We recommend providing the MeiliSearch public key, which is enough to perform searches.<br>

Reviewer: ```suggestion
**WARNING: We recommend providing the MeiliSearch public key, which is enough to perform search requests.<br>
```

Read more about [MeiliSearch authentication](https://docs.meilisearch.com/guides/advanced_guides/authentication.html).**

### Customization

The default behavior of this library fits perfectly for a documentation search bar, but you might need some customizations.

- [Optional parameters](#optional-parameters-) (when calling `docsSearchBar` method)
- [Styling](#styling-) (with CSS)

#### Optional parameters <!-- omit in toc -->

When calling the `docsSearchBar` method, you can add optional fields:

##### `queryHook` <!-- omit in toc -->

> `queryHook` takes a callback function as value. This function will be called on every keystroke to transform the typed keywords before querying MeiliSearch. By default, it does not do anything.

Reviewer: ```suggestion
`queryHook` takes a callback function as value. This function will be called on every keystroke to transform the typed keywords before querying MeiliSearch. By default, it does not do anything, but is the perfect place for you to add some preprocessing or custom functionality
```

Author: I'm not a big fan to add more and more text just to say "that is customization". But if you guys like that, I'm ok 🙂
PS: You forgot the final dot 😉

Reviewer: Dot fixed.

I understand, maybe it isn't necessary. For me it was a little bit weird to say 'this function does nothing by default' without writing a tiny line saying why did you put it there. But meybe the previous clarification is enough. I leave it up to you :)

Author: Hum, I finally take your side 😁

##### `transformData` <!-- omit in toc -->

> `transformData` takes a callback function as value. This function will be called on every hit before displaying them. By default, it does not do anything.

Reviewer: ```suggestion
`transformData` takes a callback function as value. This function will be called on every hit before displaying them. By default, it does not do anything, but it lets you add any post-processing around the data you received from MeiliSearch.
```

Author: Same as the previous suggestion.

$ git push --tag origin master
```

A GitHub Actions will be triggered and push the package on [npm](https://www.npmjs.com/package/docs-searchbar.js).

## Credits

Based on [Algolia DocSearch repository](https://github.com/algolia/docsearch) from [this commit](https://github.com/algolia/docsearch/commit/4c32b6f80b753f592de83351116664bf74b10297).<br>

> Due to a lot of future changes on this repository compared to the original one, we don't maintain it as an official fork.

Reviewer 2: ```suggestion
Due to a lot of future changes in this repository compared to the original one, we don't maintain it as an official fork.
```
